### PR TITLE
feat: add function image

### PR DESCRIPTION
**Pull request by SdgJlbl**
## Description
Add the docker image URI + checksum to the orchestrator

Reviewable commit by commit

Companion PR https://github.com/Substra/substra-backend/pull/739

Fixes FL-1149

## Checklist

- [ ] [changelog](../CHANGELOG.md) was updated with notable changes
- [ ] documentation was updated

### CODE REVIEW

File: server/standalone/migration/000060_add_function_image.up.sql [hunk 1]
```diff
@@ -0,0 +1,39 @@
+SELECT execute($$
+    ALTER TABLE functions
+    ADD COLUMN image_address VARCHAR(200) DEFAULT '' NOT NULL;
+
+    ALTER TABLE functions
+    RENAME COLUMN functionAddress to archive_address;
+
+    INSERT INTO addressables(storage_address, checksum) VALUES ('','');
```
Comment (ThibaultFy) [inline server/standalone/migration/000060_add_function_image.up.sql:8]: Worth checking if there is an other way to do it without adding this blanck line to addressables.

**Review by guilhem-barthes — APPROVED** [review]
Thanks for your contribution ! I don't remember why we decided to not accept `Image` to be null, but this looks like a decent solution keeping this in mind

**Comment by SdgJlbl** [comment]
> Thanks for your contribution ! I don't remember why we decided to not accept `Image` to be null, but this looks like a decent solution keeping this in mind

Having a nullable `Image` requires to alleviate some non-null constraints on the `events` table, and we thought that it was safer to keep these constraints and create an empty image instead.

**Comment by ThibaultFy** [comment]
/e2e --tests sdk --refs orchestrator=feat/add-function-image,substra-backend=feat/check-image-checksum,substra=feat/rename-function-to-archive,substrafl=feat/rename-function-to-archive

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: server/standalone/dbal/addressable.go [hunk 2]
```diff
@@ -14,6 +14,50 @@ func (d *DBAL) addAddressable(addressable *asset.Addressable) error {
 	return d.exec(stmt)
 }
 
+func (d *DBAL) getOrAddAddressable(addressable *asset.Addressable) error {
```
[the remark this answers is not in the document]
Comment (ThibaultFy) [reply server/standalone/dbal/addressable.go]: I proposed some changes in that sense. Tell me if it seems ok for you :)

File: server/standalone/migration/000060_add_function_image.up.sql (hunk 1 above)
Comment (SdgJlbl) [reply server/standalone/migration/000060_add_function_image.up.sql:8]: I think you created an issue for that?

Comment (ThibaultFy) [reply server/standalone/migration/000060_add_function_image.up.sql:8]: Mmmh not sure about that. Will do

Comment (ThibaultFy) [reply server/standalone/migration/000060_add_function_image.up.sql:8]: Woops you were right there was one.. I created a new one with a better context in the end 👼

File: lib/service/function_test.go [hunk 3]
```diff
@@ -173,8 +178,83 @@ func TestUpdateSingleExistingFunction(t *testing.T) {
...
+	service := NewFunctionService(provider)
+
+	existingFunction := &asset.Function{
+		Key:   "4c67ad88-309a-48b4-8bc4-c2e2c1a87a83",
+		Name:  "function name",
+		Owner: "owner",
+		Image: &asset.Addressable{StorageAddress: "test/storage/address", Checksum: "4c67ad88309a48b48bc4c2e2c1a87a83"},
 	}
 
+	// Not given image is equivalent to give Image: &asset.Addressable{StorageAddress: "", Checksum: ""}. Update an image with blank value will be ignored.
```
Comment (SdgJlbl) [inline lib/service/function_test.go:257]:
```suggestion
	// Not giving an image is equivalent to giving Image: &asset.Addressable{StorageAddress: "", Checksum: ""}. Update an image with blank value will be ignored.
```

File: server/standalone/dbal/function.go [hunk 4]
```diff
@@ -17,44 +17,52 @@ type sqlFunction struct {
...
 		return err
 	}
 
-	err = d.addAddressable(function.Function)
+	err = d.addAddressable(function.Archive, false)
+	if err != nil {
+		return err
+	}
+
+	// We allow conflict as the default Image with empty string as chesum and address already exists in table.
```
Comment (SdgJlbl) [inline server/standalone/dbal/function.go:56]:
```suggestion
	// We allow conflict as the default Image with empty string as checksum and address already exists in table.
```

**Review by guilhem-barthes — APPROVED** [review]
LGTM 🥇

**Pull request merged**